hamilton/plugins/jupyter_magic.py

```diff
@@ -77,159 +131,272 @@ def insert_cell_with_content():
+                config = json.loads(config_str)
+            except json.JSONDecodeError:
+                print(f"JSONDecodeError: Failed to parse `config` as JSON. Received {value}")
+                return
 
-        display_config = self.get_display_config(args)
+        # get the default values of args
+        module_name = args.module_name
+        base_builder = self.shell.user_ns[args.builder] if args.builder else driver.Builder()
+        inputs = self.shell.user_ns[args.inputs] if args.inputs else {}
+        overrides = self.shell.user_ns[args.overrides] if args.overrides else {}
+        display_config = (
+            self.shell.user_ns[args.display] if args.display not in [True, None] else {}
+        )
+
+        # Decision: write to file before trying to build and execute Driver
+        # See argument `help` for behavior details
+        if args.write_to_file:
+            if isinstance(args.write_to_file, str):
+                file_path = Path(args.write_to_file)
+            else:
+                file_path = Path(f"{module_name}.py")
+            file_path.write_text(cell)
 
-        module_object = ad_hoc_utils.create_module(cell, module_name, verbosity=args.verbosity)
+        # create_module() is preferred over module_from_source() to simplify
+        # the integration with the Hamilton UI which assumes physical Python modules
+        cell_module = ad_hoc_utils.create_module(cell, module_name)
+        self.shell.push({module_name: cell_module})
 
-        # shell.push() assign a variable in the notebook. The dictionary keys are variable name
-        self.shell.push({module_name: module_object})
+        # determine the Driver config
+        if config and base_builder.config:
+            return "AssertionError: Received a config -c/--config and a Builder -b/--builder with an existing config. Pass either one."
 
-        # shell.user_ns is a dictionary of all variables in the notebook
-        # rebuild drivers that use this module
-        if args.rebuild_drivers:
-            rebuilt_drivers = rebuild_drivers(
-                self.shell, module_name, module_object, verbosity=args.verbosity
-            )
-            self.shell.user_ns.update(rebuilt_drivers)
+        # build the Driver. the Builder is copied to avoid conflict with the user namespace
+        builder = base_builder.copy()
+        dr = builder.with_config(config).with_modules(cell_module).build()
 
-        # create a driver to display things for every cell with %%with_functions
-        dr = driver.Builder().with_modules(module_object).with_config(display_config).build()
-        self.shell.push({f"{module_name}_dr": dr})
+        # determine final vars
+        if args.execute not in [True, None]:
+            final_vars = self.shell.user_ns[args.execute]
+        else:
+            nodes = [n for n in dr.list_available_variables() if not n.is_external_input]
+            final_vars = topological_sort(nodes)
```

Comment:
why a topological sort? why not just sort it alphabetically?

Reply:
The output display currently doesn't print the keys, so topological sorts makes the most sense to print values

![image](https://github.com/DAGWorks-Inc/hamilton/assets/68975210/309f6d1b-6934-4f27-b739-a41f9373b8a5)